### PR TITLE
gh-108996: fix and enable test_msvcrt

### PR DESCRIPTION
**Pull request by aisk**
More details: https://github.com/python/cpython/pull/109004

* Issue: gh-108996

### CODE REVIEW

File: Lib/test/test_msvcrt.py [hunk 1]
```diff
@@ -1,9 +1,8 @@
+import ctypes
```
Comment (vstinner) [inline Lib/test/test_msvcrt.py]: ctypes is not always available. Please use ``try: import ctypes except ImportError: ctypes = None``, and then skip tests using ctypes if ctypes is None.

File: Lib/test/test_msvcrt.py [hunk 2]
```diff
@@ -64,35 +63,43 @@ def test_get_osfhandle(self):
...
-        old_stdin = sys.stdin
-        try:
-            sys.stdin = stdin
-            write_input(stdin.buffer.raw, c_encoded)
-            self.assertEqual(msvcrt.getwch(), c)
-        finally:
-            sys.stdin = old_stdin
+        with open('CONIN$', 'rb', buffering=0) as stdin:
+            h = msvcrt.get_osfhandle(stdin.fileno())
+            ctypes.windll.kernel32.FlushConsoleInputBuffer(h)
```
Comment (vstinner) [inline Lib/test/test_msvcrt.py]: Maybe it should be exposed in msvcrt, maybe as a private function just for testing.

**Review by vstinner — COMMENTED** [review]
I would prefer to start easy: reenable the whole file, but still skip tests known to fail, and then in a following PR, fix these tests. So we can spend more time to test in length the second PR.

**Comment by aisk** [comment]
@vstinner Current implementation doesn't failes in latest two runs, maybe it's been fixed. But the freebsd tests failed:

```
1 test failed again:
    test_msvcrt
```

I think it's weird, the tests should not run on freebsd.

**Comment by vstinner** [comment]
> I think it's weird, the tests should not run on freebsd.

The test failed because ctypes is not available on this FreeBSD CI:

> ModuleNotFoundError: No module named '_ctypes'

I told you, in tests, you should tolerate that ctypes is not available.

Or write a wrapper to FlushConsoleInputBuffer() in ``msvcrt`` or in ``_testconsole`` module (``PC/_testconsole.c``).

**Comment by aisk** [comment]
@vstinner Ah sorry, just understood what you mean above...

**Comment by aisk** [comment]
This change passed the tests in the last three times, I think it's ready to review.

**Comment by aisk** [comment]
Latest run failed because of freebsd test instance out of limit, not related to this change.

File: Lib/test/test_msvcrt.py [hunk 3]
```diff
@@ -64,35 +62,43 @@ def test_get_osfhandle(self):
...
-            self.assertEqual(msvcrt.getwche(), c)
-        finally:
-            sys.stdin = old_stdin
+        with open('CONIN$', 'rb', buffering=0) as stdin:
+            h = msvcrt.get_osfhandle(stdin.fileno())
+            flush_console_input_buffer(h)
+
+            old_stdin = sys.stdin
+            try:
+                sys.stdin = stdin
```
Comment (zooba) [inline Lib/test/test_msvcrt.py]: Updating `sys.stdin` shouldn't be necessary - the msvcrt.getwche() function uses the native stdin handle, not the Python object, and `write_input` takes the handle as an argument.

Comment (aisk) [reply Lib/test/test_msvcrt.py]: Thanks for the review, updated.

**Comment by zooba** [comment]
I don't have any further feedback. @vstinner are you happy with this?

**Comment by vstinner** [comment]
!buildbot Windows

**Comment by vstinner** [comment]
!buildbot AMD64 Windows

**Comment by vstinner** [comment]
> @vstinner are you happy with this?

Let's see if buildbots are happy :-)

**Comment by vstinner** [comment]
Tests passed on these 4 Windows buildbots: good.

> buildbot/AMD64 Windows11 Bigmem PR — Build done.
> buildbot/AMD64 Windows11 Non-Debug PR — Build done.
> buildbot/ARM64 Windows Non-Debug PR — Build done.
> buildbot/ARM64 Windows PR — Build done.

Example on ARM64 Windows Non-Debug PR:

```
0:06:33 load avg: 0.13 [382/463/1] test_msvcrt passed
```

**Comment by vstinner** [comment]
Merged,  thanks @aisk for your contribution. Let's see how it goes this time :-)

I'm not surprised that these tests failed. It's hard to write reliable tests on a terminal :-( Calling FlushConsoleInputBuffer() seems to make tests more reliable: good!

I enhanced/completed the commit message.

**Comment by aisk** [comment]
Thanks for the review!

**Pull request merged**